src/storage/client/StorageClient.cpp

```diff
@@ -454,5 +491,58 @@ StorageClient::get(GraphSpaceID space,
                               });
 }
 
+StatusOr<std::unordered_map<std::pair<GraphSpaceID, PartitionID>, HostAddr>>
+StorageClient::preHeartLeaders() {
+    auto fSpaces = client_->listSpaces();
+    return std::move(fSpaces).thenValue([](auto&& spaces){
+        return std::move(spaces);
+    }).thenError([](auto&& e){
+        LOG(ERROR) << "Get spaces failed: " << e.what();
+        return StatusOr<std::vector<meta::SpaceIdName>>();
+    }).thenValue([this](auto&& spaces){
+        std::vector<folly::SemiFuture<
+            StorageRpcResponse<std::pair<HostAddr, cpp2::GetLeaderResp>>>> leaders;
+        if (UNLIKELY(!spaces.ok())) {
+            LOG(WARNING) << "Get spaces failed!";
+            return folly::collect(leaders.begin(), leaders.end());
+        }
+        leaders.reserve(spaces.value().size());
+        LOG(INFO) << "Spaces size " << spaces.value().size();
+        for (auto& s : spaces.value()) {
```

Comment:
Maybe you don't need to getSpaceLeaders for every space, it will get all leaders from all spaces.